bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/BusinessPartnerService.kt

```diff
@@ -49,110 +50,137 @@ class BusinessPartnerService(
     }
 
-    private fun insertBusinessPartnerOutput(dto: BusinessPartnerOutputRequest): BusinessPartner {
-        val newEntity = businessPartnerMappings.toBusinessPartnerOutput(dto, StageType.Output, null, null)
-        return businessPartnerRepository.save(newEntity)
-            .also {
-                saveChangelog(dto.externalId, ChangelogType.CREATE, StageType.Output, checkBusinessPartnerType(dto.postalAddress.addressType))
-                initSharingState(dto)
-            }
+    private fun saveChangelog(upsertPartition: UpsertPartition) {
+        upsertPartition.existingEntities.forEach { entity -> saveChangelog(entity, ChangelogType.UPDATE) }
+        upsertPartition.newEntities.forEach { entity -> saveChangelog(entity, ChangelogType.CREATE) }
     }
 
-    private fun updateBusinessPartnerOutput(existingEntity: BusinessPartner, dto: BusinessPartnerOutputRequest): BusinessPartner {
-        businessPartnerMappings.updateBusinessPartnerOutput(existingEntity, dto, null, null)
-        return businessPartnerRepository.save(existingEntity)
-            .also {
-                saveChangelog(dto.externalId, ChangelogType.UPDATE, StageType.Output, checkBusinessPartnerType(dto.postalAddress.addressType))
-            }
-    }
+    private fun saveChangelog(partner: BusinessPartner, changelogType: ChangelogType) {
+        val businessPartnerTypes = checkBusinessPartnerType(partner.postalAddress.addressType)
+            .ifEmpty { listOf(null) }
 
-    fun getBusinessPartnersInput(pageRequest: PageRequest, externalIds: Collection<String>?): PageDto<BusinessPartnerInputDto> {
-        val stage = StageType.Input
-        val page = when {
-            externalIds.isNullOrEmpty() -> businessPartnerRepository.findByStage(stage, pageRequest)
-            else -> businessPartnerRepository.findByStageAndExternalIdIn(stage, externalIds, pageRequest)
-        }
-        return page.toPageDto(businessPartnerMappings::toBusinessPartnerInputDto)
+        businessPartnerTypes.forEach { type -> changelogRepository.save(ChangelogEntry(partner.externalId, type, changelogType, partner.stage)) }
     }
 
-    fun getBusinessPartnersOutput(pageRequest: PageRequest, externalIds: Collection<String>?): PageDto<BusinessPartnerOutputDto> {
-        val stage = StageType.Output
-        val page = when {
-            externalIds.isNullOrEmpty() -> businessPartnerRepository.findByStage(stage, pageRequest)
-            else -> businessPartnerRepository.findByStageAndExternalIdIn(stage, externalIds, pageRequest)
+    private fun assertInputStageExists(externalIds: Collection<String>) {
+        val existingExternalIds = businessPartnerRepository.findByStageAndExternalIdIn(StageType.Input, externalIds)
+            .map { it.externalId }
+            .toSet()
+
+        if (externalIds.minus(existingExternalIds).isNotEmpty()) {
+            val missing = externalIds.minus(existingExternalIds)
```

Comment:
You could move the variable out of the if block and reuse it.
Better naming: `missingExternalIds` ?

Reply:
Done